### PR TITLE
feat: Add support for constraint usage within ark_bn254

### PR DESCRIPTION
**Pull request by CPerezz**
## Description

This modifies the Cargo.toml with the following changes:

- Add ark-r1cs-std as an optional dependency.
- Add ark-curve-constraint-tests an a dev-dependency.
- Add r1cs feature which activates ark-r1cs-std depencendy.

It also adds all the necessary module files witin the constraints folder to make it work in https://github.com/arkworks-rs/curves/commit/8cb2eaef2af575f9cf46e322a880fc3f12ae56cf

closes: https://github.com/arkworks-rs/algebra/issues/719

### CODE REVIEW

File: curves/bn254/Cargo.toml [hunk 1]
```diff
@@ -16,16 +16,19 @@ edition = "2021"
...
 [dev-dependencies]
 ark-serialize = { version = "0.4.0", default-features = false }
 ark-algebra-test-templates = { version = "0.4.0", default-features = false }
 ark-algebra-bench-templates = { version = "0.4.0", default-features = false }
+ark-curve-constraint-tests = { path = "../curve-constraint-tests", default-features = false }
+ark-relations = { version = "0.4.0", default-features = false }
 
 [features]
 default = [ "curve" ]
 std = [ "ark-std/std", "ark-ff/std", "ark-ec/std" ]
```
Comment (vlopes11) [inline curves/bn254/Cargo.toml:30]: consider adding `ark-r1cs-std?/std` in case `std` feature is enabled

Comment (CPerezz) [reply curves/bn254/Cargo.toml:30]: Notice that if we don't require r1cs feature. Then loading `std` should not load this no?

I think it makes sense as it is. 
`r1cs` enables `ark-r1cs-std` and `std` enables `[ "ark-std/std", "ark-ff/std", "ark-ec/std" ]` (no r1cs unless it is explicitly required).

Comment (vlopes11) [reply curves/bn254/Cargo.toml:30]: The `?` will activate `std` for `ark-r1cs-std` only if `r1cs` is enabled

Comment (CPerezz) [reply curves/bn254/Cargo.toml]: Ohhh! I didn't know that `?` thing!! Fixed in 504a053

**Review by vlopes11 — APPROVED** [review]

**Comment by Pratyush** [comment]
Thanks for this PR! Quick question: I see that the PR adds variables for Fq and G1, but not Fq2 or G2, or for the corresponding pairing; @CPerezz would you be interested in making a followup PR that adds these components? It should be straightforward to adapt the BLS12-377 code for this. Thanks!

**Comment by CPerezz** [comment]
> Thanks for this PR! Quick question: I see that the PR adds variables for Fq and G1, but not Fq2 or G2, or for the corresponding pairing; @CPerezz would you be interested in making a followup PR that adds these components? It should be straightforward to adapt the BLS12-377 code for this. Thanks!

Sure, will try to do it within the next days. I was just interested on G1 as we wanted to verify folding decider circuits. But definitely happy to add the rest. I should have done it already.

**Pull request merged**